docs/read-more/how-to-write-a-sequence.md

```diff
@@ -0,0 +1,221 @@
+    })
+
+    input.on('end', () => {
+        this.logger.debug('Input stream ended')
+    })
+
+    // it returns an output stream
+    return out
+}
+```
+
+## Producing data (output stream)
+To stream data from a sequence we need it to return values over time. Some constructs in JavaScript that enable that are NodeJS streams, Generators and Iterables. Whatever you return from your sequence will be your ***output*** stream. We can choose whatever the solution is right for us. Here are some examples of a sequence producing a stream of integers every second.  
+Using a stream:
+```ts
+export default function() {
+    const out = new PassThrough()
+
+    let i = 0
+    setInterval(() => {
+        out.write(i++)
+    }, 1000)
+
+    return out
+}
+```
+Using an async generator:
+```ts
+export default async function*() {
+    let i = 0
+    while(true) {
+        await sleep(1000)
+        yield i++
+    }
+}
+```
+### Typescript
+Sequences that only produce data should be typed as [ReadableApp](https://hub.scramjet.org/docs/types/modules#readableapp).  
+[Here's an example](https://github.com/scramjetorg/scramjet-cloud-docs/blob/main/samples/scraping/app.ts).
+
+### Reading from output stream
+You could read this stream using our CLI, REST API, or API Client.
+
+## Consuming data (input stream)
+Reading data from a sequence is easy as the input stream conforms to the Readable protocol from NodeJS. There’s a bunch of ways that allow you to read data from streams. Here are some examples of a sequence that reads a stream of weather data objects and saves them to DB.  
+Using stream events:
+
+```ts
+export default function(input) {
+    input.on('data', (data) => {
+        saveWeatherData(data.time, data.temperature)
+    })
+}
+```
+Using a for loop:
+```ts
+export default async function*(input) {
+    for await (const data of input) {
+        saveWeatherData(data.time, data.temperature)
+    }
+}
+```
+### Input stream encoding
+Note that input stream is special in a way it already encodes your stream using a specified contentType.
+In this case if you wanted to send the stream using our CLI it would look like this:
+```bash
+si inst input <instance-id> --content-type "application/x-ndjson"
+```
+Then you can send it JSON strings separated by newline characters.  
+
+### Typescript
+Sequences that only consumes data should be typed as [WritableApp](https://hub.scramjet.org/docs/types/modules#writableapp).
+
+### Writing to input stream
+
+You can write to instance input stream using our CLI, REST API, or API Client.
+
+## Transforming data
+Transforming data is really a combination of consuming and producing, usually with some logic in between. Let’s filter the incoming input stream of numbers to include only the even ones. 
+
+Using streams:
+```ts
+export default function(input) {
+    const out = new PassThrough()
+
+    input.on('data', (num) => {
+        if(num % 2 === 0) {
+            out.write(num)
```

Comment:
Again, and downwards we need flow control

Reply:
fixed